File: taurus-protect-sdk-typescript/src/models/fee-payer.ts
/**
 * Fee payer models for Taurus-PROTECT SDK.
 *
 * Fee payers are accounts used to pay transaction fees on behalf of other
 * addresses, commonly used for sponsored transactions on EVM-compatible
 * blockchains like Ethereum.
 */

/**
 * Represents a fee payer configuration in the Taurus-PROTECT system.
 *
 * A fee payer is an account that pays transaction fees on behalf of other
 * addresses, enabling sponsored transactions on blockchains like Ethereum.
 */
export interface FeePayer {
  /** Unique fee payer identifier */
  readonly id?: string;
  /** Tenant identifier */
  readonly tenantId?: string;
  /** Blockchain type (e.g., "ETH") */
  readonly blockchain?: string;
  /** Network (e.g., "mainnet", "goerli") */
  readonly network?: string;
  /** Fee payer name */
  readonly name?: string;
  /** Creation date */
  readonly creationDate?: Date;
  /** Blockchain-specific fee payer information */
  readonly feePayerInfo?: FeePayerInfo;
}

/**
 * Blockchain-specific information for a fee payer.
 */
export interface FeePayerInfo {
  /** Blockchain type */
  readonly blockchain?: string;
  /** Ethereum-specific fee payer configuration */
  readonly eth?: FeePayerEth;
}

/**
 * Ethereum-specific fee payer configuration.
 */
export interface FeePayerEth {
  /** Kind of fee payer (e.g., "local", "remote") */
  readonly kind?: string;
  /** Local fee payer configuration */
  readonly local?: FeePayerEthLocal;
  /** Remote fee payer configuration */
  readonly remote?: FeePayerEthRemote;
  /** Encrypted remote configuration */
  readonly remoteEncrypted?: string;
}

/**
 * Local (internally managed) Ethereum fee payer configuration.
 */
export interface FeePayerEthLocal {
  /** Address ID of the fee payer account */
  readonly addressId?: string;
  /** Address ID of the forwarder contract */
  readonly forwarderAddressId?: string;
  /** Whether requests are auto-approved */
  readonly autoApprove?: boolean;
  /** Address ID of the creator account */
  readonly creatorAddressId?: string;
  /** Kind of forwarder (e.g., "OpenZeppelinForwarder") */
  readonly forwarderKind?: string;
  /** Domain separator for EIP-712 signatures */
  readonly domainSeparator?: string;
}

/**
 * Remote (externally managed) Ethereum fee payer configuration.
 */
export interface FeePayerEthRemote {
  /** URL of the remote fee payer service */
  readonly url?: string;
  /** Username for authentication */
  readonly username?: string;
  /** Address ID of the sender account */
  readonly fromAddressId?: string;
  /** Address of the forwarder contract */
  readonly forwarderAddress?: string;
  /** Address ID of the forwarder contract */
  readonly forwarderAddressId?: string;
  /** Address of the creator account */
  readonly creatorAddress?: string;
  /** Address ID of the creator account */
  readonly creatorAddressId?: string;
  /** Kind of forwarder (e.g., "OpenZeppelinForwarder") */
  readonly forwarderKind?: string;
  /** Domain separator for EIP-712 signatures */
  readonly domainSeparator?: string;
}

/**
 * Options for listing fee payers.
 */
export interface ListFeePayersOptions {
  /** Maximum number of results to return */
  limit?: number;
  /** Number of results to skip for pagination */
  offset?: number;
  /** List of specific IDs to filter by */
  ids?: string[];
  /** Blockchain to filter by */
  blockchain?: string;
  /** Network to filter by */
  network?: string;
}

File: taurus-protect-sdk-go/pkg/protect/model/fee_payer.go
package model

import "time"

// FeePayer represents a fee payer entity for managing transaction fees.
type FeePayer struct {
	// ID is the unique identifier for the fee payer.
	ID string `json:"id"`
	// TenantID is the tenant identifier.
	TenantID string `json:"tenant_id,omitempty"`
	// Blockchain is the blockchain type (e.g., "ETH").
	Blockchain string `json:"blockchain"`
	// Network is the network identifier (e.g., "mainnet", "goerli").
	Network string `json:"network,omitempty"`
	// Name is the human-readable name for the fee payer.
	Name string `json:"name,omitempty"`
	// CreationDate is when the fee payer was created.
	CreationDate time.Time `json:"creation_date"`
	// ETH contains Ethereum-specific fee payer configuration.
	ETH *FeePayerETH `json:"eth,omitempty"`
}

// FeePayerETH contains Ethereum-specific fee payer configuration.
type FeePayerETH struct {
	// Blockchain is the blockchain type within ETH config.
	Blockchain string `json:"blockchain,omitempty"`
	// Kind indicates the fee payer type (e.g., "local", "remote").
	Kind string `json:"kind,omitempty"`
	// Local contains local fee payer configuration.
	Local *ETHLocalConfig `json:"local,omitempty"`
	// Remote contains remote fee payer configuration.
	Remote *ETHRemoteConfig `json:"remote,omitempty"`
	// RemoteEncrypted contains encrypted remote configuration (base64).
	RemoteEncrypted string `json:"remote_encrypted,omitempty"`
}

// ETHLocalConfig contains configuration for local Ethereum fee payers.
type ETHLocalConfig struct {
	// AddressID is the address ID used for fee payments.
	AddressID string `json:"address_id,omitempty"`
	// ForwarderAddressID is the forwarder contract address ID.
	ForwarderAddressID string `json:"forwarder_address_id,omitempty"`
	// AutoApprove indicates if transactions are auto-approved.
	AutoApprove bool `json:"auto_approve"`
	// CreatorAddressID is the creator address ID.
	CreatorAddressID string `json:"creator_address_id,omitempty"`
	// ForwarderKind is the type of forwarder (e.g., "OpenZeppelinForwarder").
	ForwarderKind string `json:"forwarder_kind,omitempty"`
	// DomainSeparator is the EIP-712 domain separator (base64).
	DomainSeparator string `json:"domain_separator,omitempty"`
}

// ETHRemoteConfig contains configuration for remote Ethereum fee payers.
type ETHRemoteConfig struct {
	// URL is the remote fee payer service URL.
	URL string `json:"url,omitempty"`
	// Username is the authentication username.
	Username string `json:"username,omitempty"`
	// Password is the authentication password.
	Password string `json:"password,omitempty"`
	// PrivateKey is the private key for signing.
	PrivateKey string `json:"private_key,omitempty"`
	// FromAddressID is the source address ID.
	FromAddressID string `json:"from_address_id,omitempty"`
	// ForwarderAddress is the forwarder contract address.
	ForwarderAddress string `json:"forwarder_address,omitempty"`
	// ForwarderAddressID is the forwarder contract address ID.
	ForwarderAddressID string `json:"forwarder_address_id,omitempty"`
	// CreatorAddress is the creator address.
	CreatorAddress string `json:"creator_address,omitempty"`
	// CreatorAddressID is the creator address ID.
	CreatorAddressID string `json:"creator_address_id,omitempty"`
	// ForwarderKind is the type of forwarder (e.g., "OpenZeppelinForwarder").
	ForwarderKind string `json:"forwarder_kind,omitempty"`
	// DomainSeparator is the EIP-712 domain separator (base64).
	DomainSeparator string `json:"domain_separator,omitempty"`
}

// ListFeePayersOptions contains options for listing fee payers.
type ListFeePayersOptions struct {
	// Limit is the maximum number of results to return.
	Limit int64
	// Offset is the number of results to skip.
	Offset int64
	// IDs filters by specific fee payer IDs.
	IDs []string
	// Blockchain filters by blockchain type.
	Blockchain string
	// Network filters by network.
	Network string
}

// ListFeePayersResult contains the result of listing fee payers.
type ListFeePayersResult struct {
	// FeePayers is the list of fee payers.
	FeePayers []*FeePayer `json:"fee_payers"`
	// TotalItems is the total number of items available.
	TotalItems int64 `json:"total_items"`
}

// ChecksumRequest contains the request data for computing a checksum.
type ChecksumRequest struct {
	// Data is the base64-encoded data to compute checksum for.
	Data string `json:"data"`
}

// ChecksumResult contains the result of a checksum computation.
type ChecksumResult struct {
	// Checksum is the computed checksum.
	Checksum string `json:"checksum"`
}
